plugin/hooks.go

```diff
@@ -149,4 +153,16 @@ type Hooks interface {
 	// Note that this method will be called for files uploaded by plugins, including the plugin that uploaded the post.
 	// FileInfo.Size will be automatically set properly if you modify the file.
 	FileWillBeUploaded(c *Context, info *model.FileInfo, file io.Reader, output io.Writer) (*model.FileInfo, string)
+
+	// PushNotificationWillBeSent is invoked before the push notification has been sent to the push proxy server.
+	PushNotificationWillBeSent(c *Context, notification *model.PushNotification) *model.PushNotification
+
+	// PushNotificationHasBeenSent is invoked after the push notification has been sent to the push proxy server.
+	PushNotificationHasBeenSent(c *Context, notification *model.PushNotification)
+
+	// PushNotificationHasFailed is invoked after the push proxy server replies with an error.
+	PushNotificationHasFailed(c *Context, notification *model.PushNotification, err error)
+
+	// PushNotificationEnqueued is invoked just before the notification has been enqueued to be sent to the push proxy server.
+	PushNotificationEnqueued(c *Context, notificationId, notificationType, userId, channelId, postId string)
```

Comment:
Could you elaborate a bit why this is needed? Why isn't the parameter of this function `*model.PushNotification`?

Reply:
The reason behind this, is because the push notifications are enqueued in a pull of workers that are who really send the notifications to the push proxy. The notifications itself are generated by the worker, so, Before I enqueue the notification I don't have the entire notification.

Reply:
This seems to be a very "internal" hook. I can't think of a plugin (other then your notification plugin), that would use this hook. I suppose must plugin devs would just use `PushNotificationWillBeSent`.

I'm hesitant so expose a lot of hooks, because we have to support and maintain them. 1/5 on not adding this as a "public hook".

What do you think about labels this as "internal" and warn about us changing it any time? Aren't there some API endpoints where we do the same?

Reply:
I'm ok about labeling it as "internal", and do not include that in the plugins documentation (or include it but clearly explaining that is not inside the backward compatibility policy).

Reply:
Why do you need this hook?